modules/charon-core/src/main/java/org/wso2/charon3/core/objects/RoleV2.java

```diff
@@ -0,0 +1,649 @@
+                refDefinition).orElse(Collections.emptyList());
+    }
+
+    /**
+     * Set associated applications of the role.
+     */
+    public void setAssociatedApplications(List<MultiValuedComplexType> associatedApplicationsList) {
+
+        SCIMAttributeSchema complexDefinition = SCIMSchemaDefinitions.SCIMRoleSchemaDefinition.ASSOCIATED_APPLICATIONS;
+        SCIMAttributeSchema valueDefinition = SCIMSchemaDefinitions.SCIMRoleSchemaDefinition.ASC_APPLICATIONS_VALUE;
+        SCIMAttributeSchema displayDefinition = SCIMSchemaDefinitions.SCIMRoleSchemaDefinition.ASC_APPLICATIONS_DISPLAY;
+        SCIMAttributeSchema refDefinition = SCIMSchemaDefinitions.SCIMRoleSchemaDefinition.ASC_APPLICATIONS_REF;
+        addMultivaluedComplexAtribute(associatedApplicationsList, complexDefinition, valueDefinition,
+                displayDefinition, null, null, refDefinition);
+    }
+
+    /**
+     * Get the audience of the role.
+     */
+    public ComplexAttribute getAudience() {
+
+        if (!isAudienceAttributeExist()) {
+            return null;
+        }
+        return (ComplexAttribute) attributeList.get(SCIMConstants.RoleSchemaConstants.AUDIENCE);
+    }
+
+    /**
+     * Get the audience value of the role.
+     */
+    public String getAudienceValue() {
+
+        SCIMAttributeSchema audienceDefinition = SCIMSchemaDefinitions.SCIMRoleSchemaDefinition.AUDIENCE;
+        SCIMAttributeSchema audienceValueDefinition = SCIMSchemaDefinitions.SCIMRoleSchemaDefinition.AUDIENCE_VALUE;
+        return getComplexAttribute(audienceDefinition).map(aud -> getSimpleAttribute(audienceValueDefinition, aud)
+                .map(rethrowFunction(SimpleAttribute::getStringValue))
+                .orElse(null)).orElse(null);
+    }
+
+    /**
+     * Get the audience display name of the role.
+     */
+    public String getAudienceDisplayName() {
+
+        SCIMAttributeSchema audienceDefinition = SCIMSchemaDefinitions.SCIMRoleSchemaDefinition.AUDIENCE;
+        SCIMAttributeSchema audienceDisplayDefinition = SCIMSchemaDefinitions.SCIMRoleSchemaDefinition.AUDIENCE_DISPLAY;
+        return getComplexAttribute(audienceDefinition).map(aud -> getSimpleAttribute(audienceDisplayDefinition, aud)
+                .map(rethrowFunction(SimpleAttribute::getStringValue))
+                .orElse(null)).orElse(null);
+    }
+
+    /**
+     * Get the audience type of the role.
+     */
+    public String getAudienceType() {
+
+        SCIMAttributeSchema audienceDefinition = SCIMSchemaDefinitions.SCIMRoleSchemaDefinition.AUDIENCE;
+        SCIMAttributeSchema audienceTypeDefinition = SCIMSchemaDefinitions.SCIMRoleSchemaDefinition.AUDIENCE_TYPE;
+        return getComplexAttribute(audienceDefinition).map(aud -> getSimpleAttribute(audienceTypeDefinition, aud)
+                .map(rethrowFunction(SimpleAttribute::getStringValue))
+                .orElse(null)).orElse(null);
+    }
```

Comment:
can extract a private method

Reply:
fixed in [25b26c2](https://github.com/wso2/charon/pull/392/commits/25b26c2228a54f30201b37c15e189987e55ccf32)